### PR TITLE
func: init at 1.7.0

### PR DESCRIPTION
**Pull request by Maxwell-lt**
###### Description of changes

`func` is the CLI tool for building and deploying Knative functions, and works side-by-side with the main `kn` Knative CLI tool.

I'm not all that familiar with using this tool, but I was able to adapt the `kn` derivation to build it, and it seems to work. I've had some problems getting it to actually build and deploy images, but I'm assuming that's down to my configuration of podman. Additional testing might be useful.

I initialized this at the 1.7.0 version to match the current version of `kn` in nixpkgs.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by bryanasdev000** [comment]
Will review it soon :)

**Comment by bryanasdev000** [comment]
For now check maintainers line, if this is your fist contribution you may need to add yourself as a contributor to assign this package to you.

**Comment by Maxwell-lt** [comment]
Branch updated, added myself to maintainers/maintainers-list.nix.

File: pkgs/applications/networking/cluster/func/default.nix [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    rev = "knative-v${version}";
+    sha256 = "sha256-LrWRY22deh+YL/cLb+ZwK93okVPgysBoMCmo2MrbqIs=";
+  };
+
+  vendorSha256 = null;
+
+  subPackages = [ "cmd/func" ];
+
+  ldflags = [
+    "-X main.vers=v${version}"
```
Comment (bryanasdev000) [inline pkgs/applications/networking/cluster/func/default.nix:19]: Did not read the func code but you may want to add the other ldflags here per https://github.com/knative/func/blob/main/Makefile#L27, you could put date as unix epoch and for hash maybe unknown

File: pkgs/applications/networking/cluster/func/default.nix [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+  postInstall = ''
+    installShellCompletion --cmd func \
+      --bash <($out/bin/func completion bash) \
+      --zsh <($out/bin/func completion zsh)
+  '';
+
+  doInstallCheck = true;
+  installCheckPhase = ''
+    $out/bin/func version | grep ${version} > /dev/null
```
Comment (bryanasdev000) [inline pkgs/applications/networking/cluster/func/default.nix:33]: Depending on who is gonna review or merge, you may need to change this check to be like this one https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/applications/networking/cluster/kyverno/default.nix#L36

File: pkgs/top-level/all-packages.nix [hunk 3]
```diff
@@ -16773,6 +16773,8 @@ with pkgs;
 
   funnelweb = callPackage ../development/tools/literate-programming/funnelweb { };
 
+  func = callPackage ../applications/networking/cluster/func { };
```
Comment (bryanasdev000) [inline pkgs/top-level/all-packages.nix:16776]: Nitpick, but can you add it before `funnelweb`

**Review by bryanasdev000 — CHANGES_REQUESTED** [review]
Welcome aboard! Awesome work, thank you :D

Overall, LGTM but may need to fix a few nitpicks.

**Comment by bryanasdev000** [comment]
Will try to test (functionality wise) soon.

File: pkgs/applications/networking/cluster/func/default.nix (hunk 1 above)
Comment (Maxwell-lt) [reply pkgs/applications/networking/cluster/func/default.nix:19]: Added date as epoch with same format as from the makefile, set hash to version like kyverno's derivation.

File: pkgs/top-level/all-packages.nix (hunk 3 above)
Comment (Maxwell-lt) [reply pkgs/top-level/all-packages.nix:16776]: Thanks, fixed

File: pkgs/applications/networking/cluster/func/default.nix (hunk 2 above)
Comment (Maxwell-lt) [reply pkgs/applications/networking/cluster/func/default.nix]: I think I fixed this, `nix-build -A func.tests` passed. Based on the Kyverno example, I added `testers` and `func` to the input attrset of this derivation.

**Review by bryanasdev000 — APPROVED** [review]
Did a quick test with Docker on NixOS and worked like a charm.

```
cd /tmp
func create -l go hello
cd hello
func run -r local # takes a while
docker ps # in other terminal
func invoke # in other terminal
```

**Comment by bryanasdev000** [comment]
@Maxwell-lt could you please rebase at the current master?

**Comment by Maxwell-lt** [comment]
@bryanasdev000 Updated

File: pkgs/applications/networking/cluster/func/default.nix [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  passthru.tests.version = testers.testVersion {
+    package = func;
+    command = "func version";
+    version = "v${version}";
+  };
+
+  meta = with lib; {
+    description = "The Knative client library and CLI for creating, building, and deploying Knative Functions";
+    homepage = "https://github.com/knative/func";
+    changelog = "https://github.com/knative/func/releases/tag/v${version}";
```
Comment (bobby285271) [inline pkgs/applications/networking/cluster/func/default.nix:42]: https://github.com/knative/func/releases/tag/v1.7.0 gives me page not found

Comment (Maxwell-lt) [reply pkgs/applications/networking/cluster/func/default.nix]: updated to https://github.com/knative/func/releases/tag/knative-v1.7.0

File: maintainers/maintainer-list.nix [hunk 5]
```diff
@@ -8787,6 +8787,12 @@
     githubId = 4141584;
     name = "Maxime Daniel";
   };
+  maxwell-lt = {
```
[the remark this answers is not in the document]
Comment (Maxwell-lt) [reply maintainers/maintainer-list.nix:8790]: Done

**Pull request merged**